56-sphynx-cat/technical_specification.md

```diff
@@ -0,0 +1,169 @@
+
+
+### Not included:
+- Setup of a dedicated DLQ monitor or dashboard service (more on this below)
+
+
+## Additional Implementation Details:
+
+### Dead Letter Queues in Kafka
+
+![DLQ Concept](./images/dlq.png)
+
+The above illustrates the high-level concept for a Kafka DLQ. 
+
+
+![Example DLQ flow](./images/dlq_flow.png)
+
+The flow diagram above demonstrates the proposed use of a DLQ in a given service:
+
+1. Failed events are retried a configurable number of times.
+2. Upon final failure, they are published to the service-specific DLQ topic
+(the original topic name is preserved), where they await review.
+3. Events in the DLQ are manually reviewed. To ignore or reprocess the next event in the
+DLQ, the corresponding command is sent to a dedicated DLQ consumer.
+4. If the DLQ consumer is instructed to retry the event, it will publish it to a DLQ
+Retry topic, to which the main consumer actively listens.
+1. Upon consuming an event from the retry queue, the main consumer restores the original
+topic name and proceeds with the normal request flow.
+
+### Implementation of DLQ Providers in `hexkit`
+
+The work will involve the creation of a class that combines the Kafka Subscriber and the
+Kafka Publisher, which will be used in two ways.
+
+
+**`KafkaRetrySubscriber`** *(name subject to change)*  
+The primary expansion to hexkit to get the fundamental DLQ logic going. It is
+a specialized event consumer that, upon catching an unhandled error during event
+consumption, does the following:
+1. Retries the event until the configured number of retries is exhausted (using
+exponential backoff).
+   - This can reduce the likelihood of transient errors populating the DLQ.
+1. Checks auto-ignore rules to determine if the event should be dropped or published
+   - This can be implemented later when it is more clear how best to do this.
+2. Publishes the event to a service-specific DLQ topic.
+   1. The original topic should be added as a field appended to the payload.
+3. Commits the topic offsets to avoid infinitely reprocessing the failed event.
+4. Upon consuming an event from the service-specific retry topic, extracts/removes the
+original topic from the payload and proceeds with processing as if it were any other
+event (reentry).
+
+
+**`KafkaDLQSubscriber`** *(name subject to change)*  
+A specialized event consumer that will, upon instruction, consume one event from the
+configured DLQ topic and either:
+1. Ignore the event (do nothing), or
+2. Publish the event to the configured retry topic. The original topic name should still
+be contained in the payload unmodified.
+
+This class is potentially not required if requeueing is done through a 3rd party tool
+like Kafka UI. However, if requeueing is done through either the different services or
+an in-house dedicated DLQ service, then this type of consumer will be helpful.
+
+## Other: 
+
+### Potential Problems:
+
+> [!NOTE]
+> Note: The following list is by no means exhaustive: 
```

Comment:
```suggestion
> The following list is by no means exhaustive: 
```
This is superfluous with the [!NOTE] directly above

Reply:
Good eye, sniper